File: linked_list_cycle_2.py
# TC - O(n)
# SC - O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if (slow.val == fast.val):
                break
        else:
            # No cycle detected
            return None

        # Step 2: Find the start of the cycle
        # Start of the cycle by resetting one of the pointers to the head of the list and then moving both pointers one step at a time until they meet again.
        fast = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        
        return fast

File: Linked-List-cycle-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        if(! head) return NULL;
        if(!head->next) return NULL;
    
        ListNode * slow = head;
        ListNode * fast = head;
        
        while(fast && fast->next){
            
            fast = fast->next->next;
            slow = slow->next;
            
            if(fast == slow) break;
            
        }
        
        if(fast == slow){
            ListNode * temp = head;
            while(temp!=slow){
                temp = temp->next;
                slow = slow->next;
            }
            
            return slow;
        }
        
        return NULL;
        
    }
};

// Time Complexity: o(n) + o(z) where z is the start of the cycle and n is the number of nodes
// Sapce Complexity: o(1)
